pkg/k2/language_host/language_host.go

```diff
@@ -19,27 +21,47 @@ import (
 //go:embed python/python_language_host.py
 var pythonLanguageHost string
 
-type ServerAddress struct {
+type ServerState struct {
 	Log     *zap.SugaredLogger
 	Address string
-	HasAddr chan struct{}
+	Error   error
+	Done    chan struct{}
+}
+
+func NewServerState(log *zap.SugaredLogger) *ServerState {
+	return &ServerState{
+		Log:  log,
+		Done: make(chan struct{}),
+	}
 }
 
 var listenOnPattern = regexp.MustCompile(`(?m)^\s*Listening on (\S+)$`)
+var exceptionPattern = regexp.MustCompile(`(?s)(?:^|\n)\s*Exception occurred: (.+)$`)
 
-func (f *ServerAddress) Write(b []byte) (int, error) {
+func (f *ServerState) Write(b []byte) (int, error) {
 	if f.Address != "" {
 		return len(b), nil
 	}
 	s := string(b)
-	matches := listenOnPattern.FindStringSubmatch(s)
+
+	// captures the gRPC server address
+	matches := exceptionPattern.FindStringSubmatch(s)
+	if len(matches) >= 2 {
+		f.Error = errors.New(strings.TrimSpace(matches[1]))
+		f.Log.Debug(s)
+	}
+
+	// captures a fatal error in the language host that occurs before the address is printed to stdout
+	matches = listenOnPattern.FindStringSubmatch(s)
 	if len(matches) >= 2 {
 		// address is the first match
 		f.Address = matches[1]
 		f.Log.Debugf("Found language host listening on %s", f.Address)
-		close(f.HasAddr)
 	}
 
+	if f.Address != "" || f.Error != nil {
```

Comment:
I think this condition is inverted - those conditions are when it is already closed.

Reply:
If one of those is not nil/empty, we close the channel. This is the only place we close the channel. What am I missing?

Reply:
Consider the 2nd time an Error or Address are encountered, this condition would be true the second time, resulting in a double close. Unless we're doing something to shut it down, but I don't think we are (and not sure we should for cases like Address -> Error).

Reply:
Just updated this change. Now, we gate the close behind an if-else-if, so it only calls one or the other. Once Either branch evaluates to `true`, we close the channel and any subsequent calls to write will just return immediately. So you only get either an error or an address, but never both.